### PR TITLE
proxy: preserve body headers in http1

### LINKED ISSUE #447: Proxy does not handle empty request bodies correctly (except for HEAD and GET?)
This is a follow-up to #402 to track the real solution to the problem that the proxy doesn't handle requests without a body correctly. The PR for #402 only partially resolves the issue. Really we should fix the real issue. I very vaguely understand that to require some changes to the Hyper API or some other more drastic solution.

In particular, DELETE will almost never have a request body, so many REST APis would be affected by this.

/cc @olix0r @seanmonstar

### PR DESCRIPTION
**Pull request by seanmonstar**
As a goal of being a transparent proxy, we want to proxy requests and
responses with as little modification as possible. Basically, servers
and clients should see messages that look the same whether the proxy was
injected or not.

With that goal in mind, we want to make sure that body headers (things
like `Content-Length`, `Transfer-Encoding`, etc) are left alone. Prior
to this commit, we at times were changing behavior. Sometimes
`Transfer-Encoding` was added to requests, or `Content-Length: 0` may
have been removed. While RC 7230 defines that differences are
semantically the same, implementations may not handle them correctly.

Now, we've added some fixes to prevent any of these header changes
from occurring, along with tests to make sure library updates don't
regress.

For requests:

- With no message body, `Transfer-Encoding: chunked` should no longer be
added.
- With `Content-Length: 0`, the header is forwarded untouched.

For responses:

- Tests were added that responses not allowed to have bodies (to HEAD
requests, 204, 304) did not have `Transfer-Encoding` added.
- Tests that `Content-Length: 0` is preserved.
- Tests that HTTP/1.0 responses with no body headers do not have
`Transfer-Encoding` added.
- Tests that `HEAD` responses forward `Content-Length` headers (but not
an actual body).

Closes #447

### CODE REVIEW

File: proxy/tests/support/client.rs [hunk 1]
```diff
@@ -102,6 +102,7 @@ fn run(addr: SocketAddr, version: Run) -> Sender {
                     .build(&reactor);
                 Box::new(rx.for_each(move |(req, cb)| {
                     let mut req = hyper::Request::from(req.map(|()| hyper::Body::empty()));
+                    assert!(req.body_mut().take().unwrap().is_empty());
```
Comment (briansmith) [inline proxy/tests/support/client.rs]: I have no idea what this change means. I think the person who reviewed the `http` crate change needs to review this.

File: proxy/tests/transparency.rs [hunk 2]
```diff
@@ -296,13 +296,21 @@ fn http11_upgrade_not_supported() {
 }
 
 #[test]
-fn http1_get_doesnt_add_transfer_encoding() {
+fn http1_requests_without_body_doesnt_add_transfer_encoding() {
```
Comment (briansmith) [inline proxy/tests/transparency.rs]: This test looks good to me.

File: proxy/tests/support/client.rs (hunk 1 above)
Comment (seanmonstar) [reply proxy/tests/support/client.rs]: This is in the test code. The test client currently don't have any way of setting a body, so it's just asserting that it is actually empty.

**Comment by seanmonstar** [comment]
> In particular, do we rewrite a non-chunked request with a request body to use chunked encoding?

This should happen in no case:

- If there is a `Content-Length` header, then it is forwarded, and hyper understands it (and enforces it).
- If there is a `Transfer-Encoding` header, likewise, hyper forwards it.
- If there is neither, then there is no body, as this is testing.

What sort of test would you hope to see? That a `Content-Length` header is forwarded if present?

File: proxy/tests/support/client.rs (hunk 1 above)
Comment (briansmith) [reply proxy/tests/support/client.rs]: OK, gotcha. I spent too much time misunderstanding this. So, the actual bug fix was 100% in the http crate?

Comment (briansmith) [reply proxy/tests/support/client.rs]: Oh, never mind that last comment.

File: proxy/src/transparency/glue.rs [hunk 3]
```diff
@@ -81,7 +81,7 @@ impl tower_h2::Body for HttpBody {
 
     fn is_end_stream(&self) -> bool {
         match *self {
-            HttpBody::Http1(_) => false,
+            HttpBody::Http1(ref b) => b.is_empty(),
```
Comment (briansmith) [inline proxy/src/transparency/glue.rs]: Sorry, this is the change I don't understand. In the change to the `hyper` crate that adds `b.is_empty()`, which is https://github.com/hyperium/hyper/commit/2f45d5394a2f8a49442ff4798a4b1651c079f0ff, there is this comment:

```rust
    /// This does **not** detect if the body stream may be at the end, or
    /// if the stream will not yield any chunks, in all cases. For instance,
    /// a streaming body using `chunked` encoding is not able to tell if
    /// there are more chunks immediately.
```

It seems like that comment is saying that we can't implement `is_end_stream()` in terms of `b.is_empty()`, but that's what we're doing here. Should there be a similar clarifying comment here?

**Comment by briansmith** [comment]
OK, looking at this more, I think we need more test cases. Those test cases don't need to be in *this* PR, necessarily, since this PR is fixing an issue specifically with empty request bodies. However, I think the empty request body issue is actually a special case of a more general issue, which "sometimes we might transform a non-chunked request into a chunked request, but that's not transparent enough" and/or more generally "we don't handle/test all the special cases of request body forms". Is that correct, or am I still misunderstanding. If that's correct then let's at least file a follow-up issue for more complete test coverage of different request body edge cases.

File: proxy/src/transparency/glue.rs (hunk 3 above)
Comment (seanmonstar) [reply proxy/src/transparency/glue.rs]: The `Body::is_end_stream` for tower-h2 is an optimization: if it returns true, h2 won't wait on another poll, and just signal end of stream right away. So, it's OK for this hint to be **false**, as it means it was just call `poll_data` again.

The comment from hyper is trying to clarify that `Body::is_empty()` means we already know definitively that there is no chunks in this body. However, consider if the server were to receive this request:

```
POST /foo HTTP/1.1
Host: localhost
Transfer-Encoding: chunked

0


```

Since hyper saw a `Transfer-Encoding` header, it can't say for sure that there is definitely no chunks. But once the user starts streaming the body, they'll receive a `None`, since it turned out there weren't any. To know otherwise would require us to do some buffering in the proxy.

**Comment by seanmonstar** [comment]
> "sometimes we might transform a non-chunked request into a chunked request"

I don't believe that is accurate after this PR.

**Comment by seanmonstar** [comment]
@briansmith You assigned this to me, but I'm not sure what further action is needed, actually.

**Comment by briansmith** [comment]
> @briansmith You assigned this to me, but I'm not sure what further action is needed, actually.

I just set the assignments to make my issue queries clearer.

**Comment by seanmonstar** [comment]
If it's ready to merge, I need a review approval :D

**Comment by seanmonstar** [comment]
@briansmith if  there's nothing to address, mind approving, since Oliver would like your sign off too? Or if there is something to address and I misunderstood, lemme know :D

**Comment by briansmith** [comment]
I'm still struggling to understand this.

First, I still see the special-case logic for HEAD and GET in https://github.com/hyperium/hyper/blob/master/src/proto/h1/role.rs#L354-L368.

1. Why is that logic still needed when we have this new logic?
2. Why isn't the fix in this PR a generalization of the logic in that code?

Here is what I'd expect the structure of a fix for this issue to look like this (similar to the special-case logic for GET and HEAD):

```psuedo
// Avoid converting a non-chunked request into a chunked request whenever
// practical on the assumption that the underlying service can't handle chunked
// encoding. Don't bother to de-chunk smaller chunked requests, on the
// assumption that the underlying service must have already supported them if
// it inbound requests were already chunked.
let can_be_chunked = if inbound request was not chunked {
     inbound request body has a content-length and its content-length >= threshhold
} else {
    true
};

// Use can_be_chunked to decide whether to use chunked encoding can be used.
```

Since this PR is not structured that way, I have a hard time seeing whether and how it behaves equivalently to that, or how it differs from that logic. Is there no way to use this clearer structure? If not, then let's document in the code how it acheives the same effect.

**Comment by seanmonstar** [comment]
> Why is that logic still needed when we have this new logic?

It was actually added because it regressed, hyper 0.11.0 handled that exact case before (GET or HEAD without content-length). It's existence is because a user reported the regression.

As to what exactly it does: it's really just a hack to protect people who do `Request::new(Get, uri).with_body("")`. That sets the internal body to `Some(B::from(""))`. And without a `Content-Length` header, hyper knows it must be chunked (but cannot ask the body anything about it, since its generic over `B: Stream<Item=impl AsRef<[u8]>>`. 

> Why isn't the fix in this PR a generalization of the logic in that code?

Because it's hacky and wrong in other cases. The code in hyper protects two request methods which should 99.9% contain no body. But many other request methods *can* have a body without a content-length, and so it would be wrong to assume otherwise.

-----

The problem we were hitting in the proxy is this:

1. To merge the hyper server (for h1) and tower-h2 server (for h2), the requests have a unified body type, `HttpBody`. 
2. The body is always set, since we use `http::Request<B>`, which does not have an optional body.
3. The `hyper::Request<B>` has an internal `Option<B>`.
4. When converting `http::Request<B>` into `hyper::Request<B>`, since the `http` has no concept of the body not existing, means `hyper::Request` has its body set to `Some(HttpBody)`.
5. When finally passed of the hyper's client, where it no longer knows anything about the body except that it's `impl Stream`, it just notices that the body is set, and there is no `content-length`, so it **must** be chunked.

The solution here in the proxy is:

1. Just before giving the request to `hyper::Client`, check if `HttpBody::is_end_stream()`.
2. If it is, unset the body from `hyper::Request`.

Thus, with this patch, these are the **only** cases where `transfer-encoding: chunked` will be used:

1. The original incoming request has `transfer-encoding: chunked`.
2. The incoming request was from h2, claimed there are more data frames coming, and did not contain `content-length` header. However, we do not as-of-yet accept h2 and then send as h1.

File: proxy/src/transparency/client.rs [hunk 4]
```diff
@@ -174,7 +174,7 @@ where
                 let is_body_empty = req.body().is_end_stream();
                 let mut req = hyper::Request::from(req.map(BodyStream::new));
                 if is_body_empty {
-                    req.headers_mut().set(hyper::header::ContentLength(0));
+                    req.body_mut().take();
```
Comment (briansmith) [inline proxy/src/transparency/client.rs]: Let's add a test for a request like this:

```
POST /foo HTTP/1.1
Content-Length: 0
```

In particular, let's verify that the `Content-Length: 0` from the input request is preserved in the output request.

Also, please document why we're doing this (the body is a fake placeholder that is used to workaround the fact that library X doesn't directly support messages without bodies), similar to what you wrote in your recent GitHub comment.

Comment (briansmith) [reply proxy/src/transparency/client.rs]: BTW, it is confusing why we use req.body().is_end_stream()` here but then below we use `is_empty()`.

**Comment by briansmith** [comment]
Also, please rebase this PR on top of master so that #489 can be made easier to review.

File: proxy/src/transparency/client.rs (hunk 4 above)
Comment (seanmonstar) [reply proxy/src/transparency/client.rs]: The two different methods are required because the types are different. At this function, we've received an `http::Request<transparency::HttpBody>`.

Down below, `HttpBody::is_end_stream` is being implemented in terms of `Either<hyper::Body, tower_h2::RecvBody>`.

Comment (seanmonstar) [reply proxy/src/transparency/client.rs]: > In particular, let's verify that the Content-Length: 0 from the input request is preserved in the output request.

What exactly are we trying to test here? Something related to transfer-encoding?

FWIW, I don't believe the `Content-Length: 0` header will survive all the way to the end server. hyper also knows that if the body is `None` internally, and thus no body, that it can remove any set `Content-Length` and `Transfer-Encoding` headers, since a client request with neither has the same semantics.

Comment (briansmith) [reply proxy/src/transparency/client.rs]: There is a difference between an empty body (Content-Length: 0, chunked encoding of a zero-length body) and no body, and we should preserve that distinction. If we already weren't, then we can just file an issue to track fixing that. If we were already maintaining that distinction but wouldn't with this PR, then we should fix this PR to keep maintaining the distinction.

> What exactly are we trying to test here? Something related to transfer-encoding?

The test is to ensure that we don't strip the Conrtent-Length header from a request with an empty body, turning it into a request with no body.

(Rightly or wrongly, some servers treat a empty body differently than a zero-length body. For example, some versions of the HTTP spec say that a POST request must have a body, and some implementations interpret that to mean that there must always be a Content-Length or Transfer-Encoding indicating a body, even if zero-length.)

Comment (seanmonstar) [reply proxy/src/transparency/client.rs]: I see what you're saying. I haven't received a single issue filed on hyper about this case, so I do wonder how much it might happen, but I settled on the reasoning that we are a "transparent" proxy, and so not to repair or alter requests other than we absolutely have to.

As to the test, indeed, before, a `Content-Length: 0` was forwarded on, since hyper saw the set body and therefore didn't touch the headers, and after with this patch, hyper cleaned up. So, I've added a check for the existence of a `Content-Length` header, and if it exists, the body is left alone.

Also added comments explaining why this is all being done. For Future Us.

File: proxy/src/transparency/client.rs [hunk 5]
```diff
@@ -169,12 +169,32 @@ where
...
+                // we unset the body, hyper assumes the RFC7230 semantics that
+                // it can strip content body headers. Therefore, even if the
+                // body is empty, it should still be passed to hyper so that
+                // the `Content-Length: 0` header is not stripped.
                 let is_body_empty = req.body().is_end_stream();
+                let has_content_length = req.headers().contains_key(CONTENT_LENGTH);
                 let mut req = hyper::Request::from(req.map(BodyStream::new));
-                if is_body_empty {
-                    req.headers_mut().set(hyper::header::ContentLength(0));
+                if is_body_empty && !has_content_length {
```
Comment (briansmith) [inline proxy/src/transparency/client.rs]: Bigger issue: I have the same question I asked before: Why isn't this check simply `!req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING)`? If we can use `req.headers().contains_key(TRANSFER_ENCODING)` instead of `req.body().is_end_stream()` then that's much clearer to me. Otherwise, we should add a note about why we can't use a test for the presence of a Transfer-Encoding header here.

(Again, I'm hoping the code will mirror what the RFC says as closely as possible.)

File: proxy/src/transparency/client.rs [hunk 6]
```diff
@@ -169,12 +169,32 @@ where
     }
 
     fn call(&mut self, req: Self::Request) -> Self::Future {
+        use http::header::CONTENT_LENGTH;
+
         match self.inner {
             ClientServiceInner::Http1(ref h1) => {
+                // As of hyper 0.11.x, a set body implies a body must be sent.
```
Comment (briansmith) [inline proxy/src/transparency/client.rs]: Let's preface this comment with a reference to the RFC, something like:

// https://tools.ietf.org/html/rfc7230#section-3.3 says "The presence of a message body in a request is signaled by a Content-Length or Transfer-Encoding header field."

File: proxy/src/transparency/client.rs [hunk 7]
```diff
@@ -169,12 +169,32 @@ where
...
+                // want to preserve the exact intent of the original request, as
+                // we are trying to be a transparent proxy.
+                //
+                // So, if `Content-Length` is set at all, we need to send it. If
+                // we unset the body, hyper assumes the RFC7230 semantics that
+                // it can strip content body headers. Therefore, even if the
+                // body is empty, it should still be passed to hyper so that
+                // the `Content-Length: 0` header is not stripped.
                 let is_body_empty = req.body().is_end_stream();
+                let has_content_length = req.headers().contains_key(CONTENT_LENGTH);
```
Comment (briansmith) [inline proxy/src/transparency/client.rs]: NIT1: `is_body_empty` would be better named `has_no_body` since we're explicitly differentiating between a zero-length (empty) body and the lack of a body.

NIT2: Almost every request will have a content-length header set. So, it seems like it would be clearer and slightly more efficient to check if there is a Content-Length header and only check req.body().is_end_stream() if there is no Content-length set, i.e. something like:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING);
```

or:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && req.body.is_end_stream();
```

File: proxy/tests/support/client.rs [hunk 8]
```diff
@@ -102,6 +102,9 @@ fn run(addr: SocketAddr, version: Run) -> Sender {
                     .build(&reactor);
                 Box::new(rx.for_each(move |(req, cb)| {
                     let mut req = hyper::Request::from(req.map(|()| hyper::Body::empty()));
+                    if !req.headers().has::<hyper::header::ContentLength>() {
+                        assert!(req.body_mut().take().unwrap().is_empty());
```
Comment (briansmith) [inline proxy/tests/support/client.rs]: I think this should be:

```rust
assert_eq!(!req.body_mut().take().unwrap().is_empty(), req.headers().has::<hyper::header::ContentLength>());
```

Shouldn't we be considering the presence of Transfer-Encoding as an alternative to Content-Length here too?

File: proxy/tests/transparency.rs [hunk 9]
```diff
@@ -311,5 +319,68 @@ fn http1_get_doesnt_add_transfer_encoding() {
...
+    for &method in methods {
+        let resp = client.request(
+            client
+                .request_builder("/")
+                .method(method)
+                .header("content-length", "0")
+        );
+
+        assert_eq!(resp.status(), StatusCode::OK, "method={:?}", method);
+    }
```
Comment (briansmith) [inline proxy/tests/transparency.rs]: OK, great. Can we also do this same test, but using an empty body with `Transfer-Encoding: chunked`?

Finally, could you file a follow-up issue about making the same tests, except for response bodies. That is, we should test that we preserve `Content-Length: 0` in response bodies and we shouldn't convert a non-chunked response body to a chunked one (unintentionally). (Or, add the tests to this PR, if you want.)

File: proxy/src/transparency/client.rs (hunk 7 above)
Comment (briansmith) [reply proxy/src/transparency/client.rs]: Thinking about this on the way to work today, I am pretty sure that the most conservative implementation of this is:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING) &&  req.body.is_end_stream();
```

In particular, if there is a Transfer-Encoding or Content-Length header then there is a body (maybe zero-length). If neither of those headers exist, then there *might* be a body, however if we synthesized an empty body then there isn't a body.

However, I still am not convinced that we need the `req.body.is_end_stream()` check. If the original request didn't have Transfer-Encoding or Content-Length then it doesn't have a body, because there was no way for the original request to indicate the length of the body. The only way there could be a body without a Transfer-Encoding or Content-Length  is if we synthesized a body or we removed the original Transfer-Encoding/Content-Length.

Let's say we synthesized a body. Then we need to synthesize a Content-Length for it, because we can't use Transfer-Encoding for a request, since we don't know whether the server supports Transfer-Encoding. (It might be HTTP/1.0 or an incomplete HTTP/1.1 implementation.) If we are going to synthesize a Content-Length then we need to do it here.

Let's say we removed Transfer-Encoding or Content-Length from the original request, for some reason. Then again we need to synthesize a Content-Length, because we have to conservatively assume that it had a Content-Length, and because the server might not understand Transfer-Encoding.

In other words, either there must be a Transfer-Encoding in the original request, or there must be a Content-Length, or the `req.body` must be able to tell us its length so that we can synthesize a Content-Length, or there actually isn't a body. However, whether or not there is a body can always be determined by whether or not the request we forward has Content-Length (perhaps synthesized by us) or Transfer-Encoding, right?

**Comment by seanmonstar** [comment]
OK, so I think I finally understand the underlying motivation: being as we want to be as transparent of a proxy as possible, if reasonable, we shouldn't modify headers of requests and responses. That's what I'm going to change this PR to be about. I'll add some tests around responses, and a checklist, here's what exists so far:
- [x] requests
  - [x] with no body should not gain `Content-Length` nor `Transfer-Encoding`
  - [x] with `Content-Length: 0` should pass on the `Content-Length: 0` header
- [ ] responses
  - [ ] with `Content-Length: 0` should pass on the `Content-Length: 0` header
  - [ ] not allowed to have bodies (`HEAD` responses, 204, 304, etc) should not have `Content-Length` or `Transfer-Encoding` added
  - [ ] (`HEAD` responses should pass on `Content-Length` headers if they existed)
  - [ ] HTTP/1.0 with a body and no `Content-Length` should not add `Transfer-Encoding`

There's an outstanding question: should HTTP/1.1 responses with no body headers but with a body be converted into chunked?

- **Pro**: This allows the client to reuse the connection to the proxy, even if the proxy can't reuse the connection to the server.
- **Pro**: Length delimited responses are much more reliable: responses that end when the connection closes mean its possible to get a partial response and not know. For that reason, using EOF response bodies in HTTP/1.1 is *highly discouraged*.
- **Con**: We've modified the response, which wouldn't have happened if we weren't there.
- **Con**: Possibly the client doesn't understand chunked encoding (but if it claimed HTTP/1.1, it's supposed to!). 

This question is not referring to when the client claims `HTTP/1.0`. In that case, we definitely cannot send chunked (and this case is in the checklist above).

File: proxy/src/transparency/client.rs (hunk 7 above)
Comment (seanmonstar) [reply proxy/src/transparency/client.rs]: The `Transfer-Encoding` header is an HTTP/1.1 concept, removed in HTTP2. Instead of `Transfer-Encoding`, HTTP2 just signals that the body is not complete by sending `DATA` frames with the end-of-stream flag unset. While we don't *yet* translate from HTTP2 to HTTP1, the `Body::is_end_stream()` method **already** handles this distinction. And as long as we have tests for the desired behavior, the implementation can be tweaked stylistically whenever.

As for performance, I'm certain that checking `is_end_stream` first is fastest.

- Checking explicitly for `req.headers().contains_key(TRANSFER_ENCODING)` would be redundant, because in H2, it's an illegal header and wouldn't exist, and for HTTP1, hyper has already looked it up, and if it were absent, if could return `Body::empty()`.
- Checking `Body::is_end_stream()` is a simple boolean field check, whereas looking for the `CONTENT_LENGTH` requires a hashtable lookup.

In the end, I'd offer that the important part is to worry about the behavior, which we can do by defining tests.

**Comment by briansmith** [comment]
> OK, so I think I finally understand the underlying motivation: being as we want to be as transparent of a proxy as possible, if reasonable, we shouldn't modify headers of requests and responses.

Yes, that's right. And, also, if an input request is relying on some subset of HTTP features then we shouldn't increase the subset of HTTP that is used, whenever we can reasonably avoid doing so.

>  with Content-Length: 0 should pass on the Content-Length: 0 header

More generally, requests with any Content-length should pass on the Content-Length header; zero is just a special case we need to test for.

>  not allowed to have bodies (HEAD responses, 204, 304, etc) should not have Content-Length or Transfer-Encoding added

I think if we detect that a response has a body when it shouldn't have one, we should feel free to abort the request and close the connection because that's pretty much impossible to recover from. However, we might defer an implementation of that to another issue. For now, I would say that the goal is to preserve the Content-Length/Transfer-Encoding framing for responses exactly and irrespective of the details of what the request method (HEAD, etc.) or response code (204, etc.) is.

In the case of HEAD, for example, it's not the proxy's responsibility to strip off the body, and it can't be, because there simply can't be any body; otherwise we wouldn't parse the stream correctly. (This is probably different for HTTP/2.)

>  HTTP/1.0 with a body and no Content-Length should not add Transfer-Encoding

Right, both for requests and for responses.

> There's an outstanding question: should HTTP/1.1 responses with no body headers but with a body be converted into chunked?

> Con: We've modified the response, which wouldn't have happened if we weren't there.
> Con: Possibly the client doesn't understand chunked encoding (but if it claimed HTTP/1.1, it's supposed to!).

We have to assume that if the sender didn't use chunked encoding, and they used the worst possible framing for their message, then they probably did so to work around a receiver not supporting chunked encoding. In reality, they were probably just being sloppy, but we cannot distinguish those two cases and we have to be conservative in the interest of being transparent.

Note that Conduit itself will eventually support upgrading all the traffic within the mesh to HTTP/2, and I think we can also assume that people who really care about performance are going to implement HTTP/2 or at least do reasonable things with HTTP/1, so it's not worth trying to optimize HTTP/1 beyond what they've done, other than the basic things we do like load balancing.

File: proxy/src/transparency/client.rs (hunk 7 above)
Comment (briansmith) [reply proxy/src/transparency/client.rs]: > In the end, I'd offer that the important part is to worry about the behavior, which we can do by defining tests.

I generally agree with that. I think it would be ideal if our code was structured analogously to the relevant RFC requirement. However, I understand that here it might be difficult to do that because we have to consider HTTP/2, which has its own framing.

**Comment by seanmonstar** [comment]
> Conduit itself will eventually support upgrading all the traffic within the mesh to HTTP/2

Though, with that in mind, if a client connects to the proxy with an `HTTP/1.1` request, they are announcing that they understand chunked responses. If we take the position that we are allowed to translate from HTTP1 to HTTP2, it doesn't sound different in this case to allow the response body to be much more reliable.

Nevertheless, in this specific case, hyper's server does not have any way whatsoever to send a response body without `Content-Length` or `Transfer-Encoding`, and I'm not convinced to add the ability. So, for now, responses in those circumstances will be chunked. This isn't a regression in this PR.

**Comment by briansmith** [comment]
> Though, with that in mind, if a client connects to the proxy with an HTTP/1.1 request, they are announcing that they understand chunked responses.

I think that was the intent of the standard. However, in practice HTTP/1.1 implementations are frequently partial, due to lax interop testing. In practice, many implementations claiming to support HTTP/1.1 don't actually support chunked encoding.

>  If we take the position that we are allowed to translate from HTTP1 to HTTP2, it doesn't sound different in this case to allow the response body to be much more reliable.

HTTP/2 was developed by people who explicitly sought to avoid all the workarounds needed for HTTP/1 and also sought to minimize optional features. That's why HTTP/2 people developed a test suite for the framing and other details. 

> Nevertheless, in this specific case, hyper's server does not have any way whatsoever to send a response body without Content-Length or Transfer-Encoding, and I'm not convinced to add the ability. So, for now, responses in those circumstances will be chunked. This isn't a regression in this PR.

I think that's fine, especially since the issue that this PR is addressing is really about requests. The more transparent we can get for 0.3.1, the better, but we can defer issues like that particular one, to later milestones.

**Comment by seanmonstar** [comment]
Alright, I've updated this to have tests for both requests and responses.

**Comment by olix0r** [comment]
Can you please update PR description with a brief problem & solution, as would be in the final commit?

File: proxy/tests/transparency.rs [hunk 10]
```diff
@@ -360,5 +368,242 @@ fn http1_get_doesnt_add_transfer_encoding() {
...
+            .request_builder("/")
+            .method("HEAD")
+    );
+
+    assert_eq!(resp.status(), StatusCode::OK);
+    assert!(!resp.headers().contains_key("content-length"));
+    assert!(!resp.headers().contains_key("transfer-encoding"));
+
+
+
```
Comment (briansmith) [inline proxy/tests/transparency.rs]: NIT: clean up these extra blank lines.

**Comment by seanmonstar** [comment]
Updated title, description, and smashed commits into 1 with the same description.

File: proxy/tests/transparency.rs [hunk 11]
```diff
@@ -360,5 +368,240 @@ fn http1_get_doesnt_add_transfer_encoding() {
...
+                    val.to_str()
+                        .expect("req_status_header should be ascii")
+                        .parse::<u16>()
+                        .expect("req_status_header should be numbers")
+                })
+                .unwrap_or(200);
+
+            Response::builder()
+                .status(status)
+                .body(Default::default()) // default is hyper::Body::empty()
```
Comment (briansmith) [inline proxy/tests/transparency.rs]: This is confusing. If there is no body then why not use `Body::empty()`, the default? If the default works, I think it would be clearer to use it.

File: proxy/tests/transparency.rs [hunk 12]
```diff
@@ -360,5 +368,242 @@ fn http1_get_doesnt_add_transfer_encoding() {
...
+#[test]
+fn http1_head_responses() {
+    let _ = env_logger::try_init();
+
+    let srv = server::http1()
+        .route_fn("/", move |req| {
+            assert_eq!(req.method(), "HEAD");
+            Response::builder()
+                .header("content-length", "55")
+                .body(Default::default()) // default is hyper::Body::empty()
```
[the remark this answers is not in the document]
Comment (seanmonstar) [reply proxy/tests/transparency.rs]: Hm, that comment is actually trying to say that `Default::default() == hyper::Body::empty()`. And as I write this, I cannot recall why I opted to use this with a comment instead of just `hyper::Body::empty()` directly.

Perhaps since `hyper::Body` isn't used in the http2 server cases, and so to allow the code to adapt?

[the remark this answers is not in the document]
Comment (briansmith) [reply proxy/tests/transparency.rs]: OK, I see I misread it. never mind my comments here.
